File: backend/app/services/population.py
from app.utils.constants import COGS
from app.utils.raster import sum_raster_over_geom
from shapely import wkt


def get_population_stats(aoi_wkt: str):
    geom = wkt.loads(aoi_wkt)

    totals = {key: sum_raster_over_geom(path, geom) for key, path in COGS.items()}
    totals["all"] = sum(totals.values())

    percentages = (
        {key: 0 for key in ["lt15", "age15_64", "gt65"]}
        if totals["all"] == 0
        else {
            key: round(100 * totals[key] / totals["all"], 2)
            for key in ["lt15", "age15_64", "gt65"]
        }
    )

    return {"totals": totals, "percentages": percentages}

File: frontend/src/types/population.ts
import type { GeoJSONFeature, GeoJSONGeometry } from "wellknown";

export type PopulationStatsType = {
  aoi_wkt: string;
  totals: {
    lt15: number;
    age15_64: number;
    gt65: number;
    all: number;
  };
  percentages: {
    lt15: number;
    age15_64: number;
    gt65: number;
  };
  name: string;
  country: string;
};

export type PopulationAgeParamsType = {
  feature: GeoJSONGeometry | GeoJSONFeature;
  name: string;
  country: string;
};
